src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -988,6 +1014,47 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
             return node;
         }
 
+        /// <returns>
+        /// <see langword="true"/> if the pattern only matches if the top-level input is not null.
+        /// <see langword="false"/> if the pattern only matches if the top-level input is null.
+        /// Otherwise, <see langword="null"/>.
+        /// </returns>
+        private bool? GetTopLevelNullTest(BoundPattern pattern)
+        {
+            switch (pattern)
+            {
+                case BoundTypePattern:
+                case BoundRecursivePattern:
+                case BoundITuplePattern:
+                case BoundRelationalPattern:
+                case BoundDeclarationPattern { IsVar: false }:
+                case BoundConstantPattern { ConstantValue: { IsNull: false } }:
+                    return true;
+                case BoundConstantPattern { ConstantValue: { IsNull: true } }:
+                    return false;
+                case BoundNegatedPattern negated:
+                    return !GetTopLevelNullTest(negated.Negated);
+                case BoundBinaryPattern binary:
+                    if (binary.Disjunction)
+                    {
+                        // `a?.b(out x) is null or C`
+                        // both subpatterns must have the same null test for the test to propagate out
+                        var leftNullTest = GetTopLevelNullTest(binary.Left);
+                        return leftNullTest is null ? null :
+                            leftNullTest != GetTopLevelNullTest(binary.Right) ? null :
+                            leftNullTest;
+                    }
+
+                    // `a?.b(out x) is not null and var c`
+                    // if any pattern performs a test, we know that test applies at the top level
+                    // note that if the tests are different, e.g. `null and not null`,
+                    // the pattern is a contradiction, so we expect an error diagnostic
+                    return GetTopLevelNullTest(binary.Left) ?? GetTopLevelNullTest(binary.Right);
+            }
+
+            return null;
```

Comment:
This should probably throw on unexpected patterns, not return.

Reply:
Done. Have manually enumerated the pattern kinds, which may help us catch it and decide what to do when a new pattern kind is added to the language.